pkg/virt-controller/watch/vm.go

```diff
@@ -544,6 +546,40 @@ func (c *VMController) handleDataVolumes(vm *virtv1.VirtualMachine, dataVolumes
 	return ready, nil
 }
 
+// areDataVolumesReady determines whether all DataVolumes specified for a VM
+// have been successfully provisioned and/or claimed, and are ready for consumption.
+// Note that DataVolumes in WaitForFirstConsumer phase are not regarded as ready.
+func (c *VMController) areDataVolumesReady(vm *virtv1.VirtualMachine) bool {
+	for _, template := range vm.Spec.DataVolumeTemplates {
+		dvKey := fmt.Sprintf("%s/%s", vm.Namespace, template.Name)
+		dvObj, exists, err := c.dataVolumeInformer.GetStore().GetByKey(dvKey)
+		if err != nil {
+			log.Log.Object(vm).Errorf("Error fetching DataVolume %s: %v", dvKey, err)
+			return false
+		}
+		if !exists {
+			return false
+		}
+
+		dv := dvObj.(*cdiv1.DataVolume)
+
+		if !v1.IsControlledBy(dv, vm) {
+			// If the DataVolume had just been adopted by this controller in the current reconcile cycle,
+			// this will not be detected here since the dataVolumeInformer cache isn't likely
+			// to be updated with the new ownerReference just yet. Once the cache is updated,
+			// another reconcile cycle will be triggered in which the adoption will be detected
+			// and the Provisioning status can potentially be removed (assuming all other DVs are ready).
+			return false
+		}
+
+		if dv.Status.Phase != cdiv1.Succeeded {
+			return false
+		}
+	}
+
+	return true
+}
```

Comment:
There's no rule the a DataVolume has to be defined in the DataVolumeTemplate section on a VM or that every DataVolume used by a VM is controlled by the corresponding VM

I think we can simplify this check and make it more accurate. Instead of looping through the DataVolumeTemplate list, loop through the vm.spec.template.spec.volumes list. For every DataVolume found there, find it in the cache, and simply verify that `dv.Status.Phase == cdiv1.Succeeded`

Reply:
That's a good point. I've reworked the logic to loop through `vm.spec.template.spec.volumes`.

It is, however, might be misleading, if we display a "Provisioning" status for a VM with a referenced DataVolume that doesn't exist, as there's nothing in the system that actively tries to provision that DataVolume. In that case it may be better to display a different status that highlights the fact that the system is waiting for some action (DataVolume creation) to be taken.

Reply:
And another though - we may also need to examine volumes of type persistentVolumeClaim, and verify whether they are bound (ready), in-progress (--> Provisioning status), or just don't match anything (which is quite similar to the above case with DVs).

Reply:
> If you decide to introspect PVCs, the basic check is confirming that the PVC is bound. From there if the PVC is owned by a DataVolume, you could cross reference the DataVolume to determine if the PVC is populated as well.

just to clarify though. I don't think this is really necessary. I've (lgtm) the pr, you can follow up wiht this change if you want. It doesn't matter much.

Reply:
I've indeed added the PVC-volumes check. I don't check the DV owner of the PVC, since in the first place I'm only directly reaching to PVCs which are referenced from volumes of type PersistentVolumeClaim (and DVs are reached directly for volume of type DataVolume...).